dali/pipeline/data/tensor_vector.cc

```diff
@@ -13,12 +13,163 @@
+    order = src_order ? src_order : dst_order;
+
+  // Wait on the order on which we will run the copy for the work to finish on the dst
+  order.wait(dst_order);
+
+  return order;
+}
+
+
+/**
+ * @brief Wait for the reallocation to happen in the copy order, so we can actually proceed.
+ */
+void SyncAfterResize(AccessOrder dst_order, AccessOrder copy_order) {
+  copy_order.wait(dst_order);
+}
+
+
+/**
+ * @brief Wait for the copy to finish in the order of the dst buffer.
+ */
+void SyncAfter(AccessOrder dst_order, AccessOrder copy_order) {
+  dst_order.wait(copy_order);
+}
+
+/**
+ * @brief Copy between two non-contiguous batches
+ * Assumes matching shapes and types
+ */
+template <typename DstBackend, typename SrcBackend, template <typename> typename DstBatch,
+          template <typename> typename SrcBatch>
+void CopySamplewiseImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src,
+                        const TypeInfo &type_info, AccessOrder order,
+                        bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, 256> srcs;
+  srcs.reserve(num_samples);
+  SmallVector<void *, 256> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    srcs.emplace_back(src.raw_tensor(i));
+    sizes.emplace_back(src.shape()[i].num_elements());
+  }
+
+  type_info.Copy<SrcBackend, DstBackend>(dsts.data(), srcs.data(), sizes.data(), num_samples,
+                                         order.stream(), use_copy_kernel);
+}
+
+
+/**
+ * @brief Copy to non-contiguous batch from contiguous source.
+ * Assumes matching shapes and type.
+ */
+template <typename DstBackend, typename SrcBackend, template <typename> typename DstBatch>
+void CopySamplewiseImpl(DstBatch<DstBackend> &dst, const void *src, const TypeInfo &type_info,
+                        AccessOrder order, bool use_copy_kernel = false) {
+  auto num_samples = dst.num_samples();
+  SmallVector<void *, 256> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    sizes.emplace_back(dst.shape()[i].num_elements());
+  }
+
+  type_info.Copy<DstBackend, SrcBackend>(dsts.data(), src, sizes.data(), num_samples,
+                                         order.stream(), use_copy_kernel);
+}
+
+
+/**
+ * @brief Copy from non-contiguous batch to contiguous destination.
+ * Assumes matching shapes and types.
+ */
+template <typename DstBackend, typename SrcBackend, template <typename> typename SrcBatch>
+void CopySamplewiseImpl(void *dst, const SrcBatch<SrcBackend> &src, const TypeInfo &type_info,
+                        AccessOrder order, bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, 256> srcs;
```

Comment:
Why 256?

Reply:
Hmm, this is a constant that is used for such copies throughout of our codebase. Maybe I can make it some constant that is defined once.